### ISSUE TITLE
Block bad bot traffic

### ISSUE THREAD

**Issue by alexroper** [FDPA/fdpa#193]
Checking Craft's logs I found a bunch of suspicious 404 requests. Not sure how but we're already getting some junk request to the server even tho we're not on the live domain yet.

Here's an example of the 404s we're getting:

```
REQUEST_URI=/phpMyAdmin-2/scripts/setup.php
REQUEST_URI=/mysql/scripts/setup.php
REQUEST_URI=/cgi/common.cgi
REQUEST_URI=/stssys.htm
REQUEST_URI=/command.php
REQUEST_URI=/etc/lib/pChart2/examples/imageMap/index.php
REQUEST_URI=/w00tw00t.at.blackhats.romanian.anti-sec:)
REQUEST_URI=/phpMyAdmin/scripts/setup.php
REQUEST_URI=/phpmyadmin/scripts/setup.php
```

Full log is attached: [404s.log.zip](https://github.com/FDPA/fdpa/files/1028175/404s.log.zip)

This is obviously bot traffic trying to find scripts to try and exploit.

Is there a way we can blacklist this traffic and other known bad bot IP addresses?

**Comment by sarawb** [comment]
@alexroper was talking about this with @wnstn and he wasn't sure about what to do for this. Any more info/advice?

**Comment by alexroper** [comment]
Hmmm. I'm not totally sure what the best way to do this is.

Something like CloudFlare has this kind of thing built in, but the client would have to spend $20 a month for it. They block potentially bad traffic and use CAPTCHAs to let thru users that show up as false positives.

We could try to just find the IP addresses, referrers or user agent strings used by the bots. But we might end up blocking legitimate IP addresses and users in the process. It might not be worth it. It's not a whole lot of traffic and they are trying to exploit weaknesses that aren't there (like Wordpress logins).

This might be a bigger issue if it's polluting Google Analytics data.

I found a couple this about this might be worth reading:

https://moz.com/blog/how-to-stop-spam-bots-from-ruining-your-analytics-referral-data

https://www.optimizesmart.com/geek-guide-removing-referrer-spam-google-analytics/

**Comment by alexroper** [comment]
Okay, I did some more research. We could do this manually by combing thru 404 logs and making lists of bad traffic and blocking it. Some people even go as far as blocking large ranges of IP addresses like from a country they don't want or care about getting traffic from. Usually China.

But that's pretty labor intensive and probably not very future proof. 

A couple automatic options would be CloudFlare or fail2ban.

CloudFlare has a free plan that let's you use their DDoS protection and "Threat Blocking & Privacy" features.

> **Block or Challenge Visitors**
> Fully block or challenge visitors by country, IP address, or AS number. Outside of the Enterprise plan, you can only 'Challenge' visitors by two letter country code, which means a human visitor could still enter the site by passing the captcha.

That's from their pricing page under the security section:
https://www.cloudflare.com/plans/pro/#compare-features

Fail2ban is free software you install on the server, not a subscription based service. It's mostly used to protect against SSH, FTP and SFTP brute force password hacking.

https://www.digitalocean.com/community/tutorials/how-to-protect-an-apache-server-with-fail2ban-on-ubuntu-14-04

>When operating a web server, it is important to implement security measures to protect your site and users. Protecting your web sites and applications with firewall policies and restricting access to certain areas with password authentication is a great starting point to securing your system. However, any publicly accessible password prompt is likely to attract brute force attempts from malicious users and bots.

>Setting up fail2ban can help alleviate this problem. When users repeatedly fail to authenticate to a service (or engage in other suspicious activity), fail2ban can issue a temporary bans on the offending IP address by dynamically modifying the running firewall policy. Each fail2ban "jail" operates by checking the logs written by a service for patterns which indicate failed attempts. Setting up fail2ban to monitor Apache logs is easy using the included configuration filters.

This software requires a lot of configuration tho. And I'm not comfortable setting something like this up. Apparently it's possible that you can unwittingly lock yourself out of the server with too many incorrect password attempts.

**Comment by sarawb** [comment]
It seems like Cloudflare is a good bet and I like that DO recommends it. @alexroper is that something you'd have time to set up? Or @wnstn?

**Comment by alexroper** [comment]
Yes, I think I can set this up. First we'll need to create an account for FDPA. Can you do this and send me the login?

In CloudFlare's setup guide they recommend using an email alias:

> Pro Tip: We recommend using an email alias or distribution list. For example, cloudflare@company.com. Billing and service-related email notifications will be sent to this email address.

https://support.cloudflare.com/hc/en-us/articles/201720164-Step-2-Create-a-Cloudflare-account-and-add-a-website

**Comment by sarawb** [comment]
I can definitely set up an account, but I am not sure I can get an email alias set up for them quite as easily. Let me ask.
